src/eval/operation.rs

```diff
@@ -1115,6 +1108,110 @@ fn process_unary_operation(
+                                    mk_term::op1(UnaryOp::Seq(), mk_term::op1(UnaryOp::Force(), t)),
+                                    acc
+                                )
+                                .with_pos(pos_op_inh)
+                            },
+                        ),
+                    ),
+                    cont
+                )
+            }
+
+            fn seq_terms<I>(mut it: I, pos_op_inh: TermPos, cont: RichTerm) -> RichTerm
+            where
+                I: Iterator<Item = RichTerm>,
+            {
+                it.fold(cont, |acc, t| mk_app!(mk_term::op1(UnaryOp::Seq(), t), acc))
+            }
+
+            // FIXME: avoid this nonsense.
+            let st = t.clone();
+
+            match t.into_owned() {
+                Term::Record(map, attrs) if !map.is_empty() => {
+                    let mut shared_env = env.clone();
+                    let map = map
+                        .into_iter()
+                        .map(|(id, t)| {
+                            (
+                                id,
+                                mk_term::op1(UnaryOp::Force(), t)
+                                    .closurize(&mut shared_env, env.clone()),
+                            )
+                        })
+                        .collect::<std::collections::HashMap<_, _>>();
+
+                    // FIXME: avoid all the cloning.
+                    let terms = map.clone().into_values();
+                    let record = RichTerm::new(Term::Record(map, attrs), pos.into_inherited());
+
+                    Ok(Closure {
+                        body: seq_terms(terms, pos_op, record),
+                        env: shared_env,
+                    })
+                }
+                Term::Array(ts, attrs) if !ts.is_empty() => {
+                    if !attrs.applied_contracts {
+                        let mut shared_env = env.clone();
+                        let terms_with_ctrs = ts
+                            .iter()
+                            .cloned()
+                            .map(|t| {
+                                apply_contracts(
+                                    t,
+                                    attrs.pending_contracts.iter().cloned(),
+                                    pos.into_inherited(),
+                                )
+                                .closurize(&mut shared_env, env.clone())
+                            })
+                            .collect();
+
+                        let array_with_ctrs = mk_term::op1(
+                            UnaryOp::Force(),
+                            RichTerm::new(
+                                Term::Array(terms_with_ctrs, attrs.applied_contracts(true)),
+                                pos,
+                            ),
+                        );
+
+                        Ok(Closure {
+                            body: force_terms(ts.into_iter(), pos_op, array_with_ctrs),
```

Comment:
If I understand correctly, the behavior changed because `Ignore` is not applied during `deep_seq`?

What's bugging me a bit is now we have two "different" ways of doing essentially the same thing (like we discussed a the beginning): `seq (force x) y` and `deep_seq x y`. I fear that people would think of these as equivalent. Afterall, the Haskell deepseq package simply defines `force x = deepSeq x x`.

Reply:
The behavior changes because right now `Ignore` drop its parameter, making it not evaluated, but the new version start by forcing the parameter anyway.

I propose to have the same behavior for both `force` and `deep_seq`, which is
 to not force the parameter of a contract. Doing so `seq (force x) y` and `deep_seq x y` should have strictly the same behavior. The only two that would be different are `force x` and `deep_seq x x`, whose return value and effect on `x` may be different. We can also define `deep_seq` as `seq (force x) y` in which case the only difference would be the returned value.

Reply:
Okay. Right now `deep_seq` evaluated the applied contracts. And so does `force` with respect to Arrays.

I think we can redefine `deep_seq` in a separate PR after merging this.